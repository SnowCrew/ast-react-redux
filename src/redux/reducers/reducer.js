

const initialState = {
  products: [],
  loading: true,
  error: false,
  itemsInCart: [],
  total: 0
};

const reducer = (state = initialState, action) => {
  switch (action.type) {
    case 'MENU_LOADED':
      return {
        ...state,
        products: action.payload,
        loading: false,
        error: false
      };
    case 'MENU_REQUESTED':
      return {
        ...state,
        products: state.products,
        loading: true,
        error: false
      };
    case 'MENU_ERROR':
      return {
        ...state,
        products: state.products,
        loading: true,
        error: true
      };
    case "ITEM_ADD_TO_CART":
      const id = action.payload;

      const itemInd = state.itemsInCart.findIndex(item => item.id === id);

      if (itemInd >= 0) {
        const elemInState = state.itemsInCart.find(item => item.id === id);
        const newElem = {
          ...elemInState,
          counterIdentical: ++elemInState.counterIdentical
        }
        return {
          ...state,
          itemsInCart: [
            ...state.itemsInCart.slice(0, itemInd),
            newElem,
            ...state.itemsInCart.slice(itemInd + 1)
          ],
          total: (Number(state.total) + Number(newElem.price)).toFixed(2)
        }
      }

      const item = state.products.find(item => item.id === id);
      const newItem = {
        title: item.title,
        price: item.price,
        image: item.image,
        id: item.id,
        counterIdentical: 1,
      }
      return {
        ...state,
        itemsInCart: [
          ...state.itemsInCart,
          newItem
        ],
        total: (Number(state.total) + Number(newItem.price)).toFixed(2)
      };

    case "ITEM_REMOVE_FROM_CART":
      const idx = action.payload;
      const itemIndex = state.itemsInCart.findIndex(item => item.id === idx);
      const itemDel = state.itemsInCart.find(item => item.id === idx);
      return {
        ...state,
        itemsInCart: [
          ...state.itemsInCart.slice(0, itemIndex),
          ...state.itemsInCart.slice(itemIndex + 1)
        ],
        total: state.total - (itemDel.price * itemDel.counterIdentical)
      };

    default:
      return state;
  }
}

export default reducer;