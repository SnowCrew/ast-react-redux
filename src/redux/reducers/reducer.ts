import { IAction } from '../actions/actions';


const initialState:IStore = {
  products: [],
  loading: true,
  error: false,
  itemsInCart: [],
  total: 0,
  totalQuantity: 0,
  auth: false,
  admin: false
};

export interface IProduct {
  id: number,
  title:string,
  price:number,
  category:string,
  description:string,
  image:string
}

export interface IProductInCart extends IProduct {
counterIdentical:number,
}

export interface IStore {
  products: IProduct[],
  loading: boolean,
  error: boolean,
  itemsInCart: IProductInCart[],
  total: number,
  totalQuantity: number,
  auth: boolean,
  admin: boolean
};

const reducer = (state:IStore = initialState, action:IAction):IStore => {
  switch (action.type) {
    case 'MENU_LOADED':
      return {
        ...state,
        products: action.payload as IProduct[],
        loading: false,
        error: false
      };
    case 'MENU_REQUESTED':
      return {
        ...state,
        products: state.products,
        loading: true,
        error: false
      };
    case 'MENU_ERROR':
      return {
        ...state,
        products: state.products,
        loading: true,
        error: true
      };
    case "ITEM_ADD_TO_CART":
      const [id, quantity] = action.payload as [number, number];
      const itemInd = state.itemsInCart?.findIndex(item => item.id === id);

      if (itemInd >= 0 ) {
        const elemInState = state.itemsInCart?.find(item => item.id === id);
        const newElem = {
          ...elemInState,
          counterIdentical: (Number(elemInState?.counterIdentical) + Number(quantity)),
        }
        return {
          ...state,
          itemsInCart: [
            ...state.itemsInCart!.slice(0, itemInd),
            newElem,
            ...state.itemsInCart!.slice(itemInd! + 1)
          ] as IProductInCart[],
          total: Number((Number(state.total) + (Number(newElem.price) * Number(quantity))).toFixed(2)),
          totalQuantity: Number(state.totalQuantity) + Number(quantity)
        }
      }

      const item = state.products.find(item => item.id === id);
      const newItem:IProductInCart = {
        title: item!.title,
        price: item!.price,
        image: item!.image,
        id: item!.id,
        category:item!.category,
        description: item!.description,
        counterIdentical: Number(quantity)
      }
      return {
        ...state,
        itemsInCart: [
          ...state.itemsInCart,
          newItem
        ],
        total: Number(((state.total) + (Number(newItem.price) * Number(quantity))).toFixed(2)),
        totalQuantity: Number(state.totalQuantity) + Number(quantity)
      };

    case "ITEM_REMOVE_FROM_CART":
      const idx = action.payload as number;

      const itemIndex = state.itemsInCart.findIndex(item => item.id === idx);
      const itemDel = state.itemsInCart.find(item => item.id === idx);
      if(itemIndex === -1){
        return state;
      }
      return {
        ...state,
        itemsInCart: [
          ...state.itemsInCart.slice(0, itemIndex),
          ...state.itemsInCart.slice(itemIndex + 1)
        ],
        total: Number((state.total - (itemDel!.price * itemDel!.counterIdentical)).toFixed(2)),
        totalQuantity: Number(state.totalQuantity) - itemDel!.counterIdentical
      };

    case "CLEAR_CART":
      return {
        ...state,
        itemsInCart: [],
        total: 0,
        totalQuantity: 0
      }

    case "GET_AUTH_REQUEST":
      return {
        ...state,
        loading: true
      };

    case "GET_AUTH":
      return {
        ...state,
        auth: true,
        loading: false
      };

    case "GET_ADMIN_AUTH":
      return {
        ...state,
        auth: true,
        loading: false,
        admin: true
      };

    case "GET_AUTH_REJECT":
      return {
        ...state,
        loading: false
      };

    case "EXIT_AUTH":
      return {
        ...state,
        auth: false,
        admin: false
      };

    default:
      return state;
  };
};

export default reducer;