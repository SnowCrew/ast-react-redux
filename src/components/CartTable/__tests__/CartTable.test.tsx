import { cleanup, render, screen } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import { ReactElement } from "react";
import { Provider } from "react-redux";
import { legacy_createStore as createStore } from "redux";
import reducer from "../../../redux/reducers/reducer";
import CartTableComponent from "../CartTable.component";

const renderWithRedux = (
  component: ReactElement,
  { initialState, store = createStore(reducer, initialState) } = {}
) => {
  return {
    ...render(<Provider store={store}>{component}</Provider>),
  };
};

describe("CartTable component", () => {
  afterEach(cleanup);
  describe("Render CartTable with initial state", () => {
    afterEach(cleanup);

    it("Render CartTable with initial state", () => {
      expect(renderWithRedux(<CartTableComponent />)).toMatchSnapshot();
      renderWithRedux(<CartTableComponent />);
      expect(screen.getAllByText("Ваша корзина пуста!")[0].textContent).toBe(
        "Ваша корзина пуста!"
      );
    });
  });
  describe("Testing CartTable with items in Cart", () => {
    afterEach(cleanup);
    it("Render CartTable with items", () => {
      expect(
        renderWithRedux(<CartTableComponent />, {
          initialState: {
            products: [],
            loading: false,
            error: false,
            itemsInCart: [
              {
                id: 1,
                title: "string",
                price: 2,
                category: "string",
                description: "string",
                image: "string",
                counterIdentical: 1,
              },
              {
                id: 2,
                title: "string2",
                price: 1,
                category: "string2",
                description: "string2",
                image: "string2",
                counterIdentical: 2,
              },
            ],
            total: 3,
            totalQuantity: 3,
            auth: false,
            admin: false,
          },
        })
      ).toMatchSnapshot();
    });

    it("CartTable handlers testing", () => {
      renderWithRedux(<CartTableComponent />, {
        initialState: {
          products: [],
          loading: false,
          error: false,
          itemsInCart: [
            {
              id: 1,
              title: "string",
              price: 2,
              category: "string",
              description: "string",
              image: "string",
              counterIdentical: 1,
            },
            {
              id: 2,
              title: "string2",
              price: 1,
              category: "string2",
              description: "string2",
              image: "string2",
              counterIdentical: 2,
            },
          ],
          total: 3,
          totalQuantity: 3,
          auth: false,
          admin: false,
        },
      });

      //add item №1 quantity +
      userEvent.click(screen.getAllByText("+")[0]);
      expect(screen.getAllByTestId("item-counter-count")[0].textContent).toBe(
        "2"
      );
      userEvent.click(screen.getAllByText("+")[1]);
      expect(screen.getAllByTestId("item-counter-count")[1].textContent).toBe(
        "3"
      );

      //decrease item quantity -
      userEvent.click(screen.getAllByTestId("decrease_button")[0]);
      userEvent.click(screen.getAllByTestId("decrease_button")[0]);
      expect(screen.getAllByTestId("item-counter-count").length).toBe(1);

      //delete item from Cart
      userEvent.click(screen.getAllByTestId("remove_button")[0]);
      expect(screen.getByText("Ваша корзина пуста!").textContent).toBe(
        "Ваша корзина пуста!"
      );
    });

    it("CartTable clear cart handler test", () => {
      renderWithRedux(<CartTableComponent />, {
        initialState: {
          products: [],
          loading: false,
          error: false,
          itemsInCart: [
            {
              id: 1,
              title: "string",
              price: 2,
              category: "string",
              description: "string",
              image: "string",
              counterIdentical: 1,
            },
            {
              id: 2,
              title: "string2",
              price: 1,
              category: "string2",
              description: "string2",
              image: "string2",
              counterIdentical: 2,
            },
          ],
          total: 3,
          totalQuantity: 3,
          auth: false,
          admin: false,
        },
      });
      //Clear Cart
      userEvent.click(screen.getAllByTestId("clear_cart_button")[0]);
      expect(screen.getByText("Ваша корзина пуста!").textContent).toBe(
        "Ваша корзина пуста!"
      );
    });
  });
});
