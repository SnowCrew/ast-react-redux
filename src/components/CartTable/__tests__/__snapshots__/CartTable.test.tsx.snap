// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CartTable component Render CartTable with initial state Render CartTable with initial state 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="cart__title"
      >
        Ваша корзина пуста!
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="cart__title"
    >
      Ваша корзина пуста!
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`CartTable component Testing CartTable with items in Cart Render CartTable with items 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="cart"
      >
        <div
          class="cart__title"
        >
          Ваш заказ:
        </div>
        <div
          class="cart__list"
        >
          <div
            class="cart__item"
          >
            <div>
              1
            </div>
            <div
              class="cart__item-title"
            >
              string
            </div>
            <div
              class="cart__item-price"
            >
              2
              $
            </div>
            <div
              class="cart__item-counter"
            >
              <button
                class="cart__item-counter-btn"
              >
                +
              </button>
              <div
                class="cart__item-counter-count"
                data-testid="item-counter-count"
              >
                1
              </div>
              <button
                class="cart__item-counter-btn minus"
                data-testid="decrease_button"
              >
                -
              </button>
            </div>
            <div
              class="cart__item-price"
              data-testid="total-price-item-counter-count"
            >
              Total: 
              2.00
              $
            </div>
            <div
              class="cart__close"
              data-testid="remove_button"
            >
              ×
            </div>
          </div>
          <div
            class="cart__item"
          >
            <div>
              2
            </div>
            <div
              class="cart__item-title"
            >
              string2
            </div>
            <div
              class="cart__item-price"
            >
              1
              $
            </div>
            <div
              class="cart__item-counter"
            >
              <button
                class="cart__item-counter-btn"
              >
                +
              </button>
              <div
                class="cart__item-counter-count"
                data-testid="item-counter-count"
              >
                2
              </div>
              <button
                class="cart__item-counter-btn minus"
                data-testid="decrease_button"
              >
                -
              </button>
            </div>
            <div
              class="cart__item-price"
              data-testid="total-price-item-counter-count"
            >
              Total: 
              2.00
              $
            </div>
            <div
              class="cart__close"
              data-testid="remove_button"
            >
              ×
            </div>
          </div>
          <div
            class="cart-total-money"
          >
            Сумма за все товары : 
            3
            $
          </div>
          <button
            class="cart__btn"
            disabled=""
          >
            Оплатить
          </button>
          <button
            data-testid="clear_cart_button"
          >
            Очистить корзину
          </button>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="cart"
    >
      <div
        class="cart__title"
      >
        Ваш заказ:
      </div>
      <div
        class="cart__list"
      >
        <div
          class="cart__item"
        >
          <div>
            1
          </div>
          <div
            class="cart__item-title"
          >
            string
          </div>
          <div
            class="cart__item-price"
          >
            2
            $
          </div>
          <div
            class="cart__item-counter"
          >
            <button
              class="cart__item-counter-btn"
            >
              +
            </button>
            <div
              class="cart__item-counter-count"
              data-testid="item-counter-count"
            >
              1
            </div>
            <button
              class="cart__item-counter-btn minus"
              data-testid="decrease_button"
            >
              -
            </button>
          </div>
          <div
            class="cart__item-price"
            data-testid="total-price-item-counter-count"
          >
            Total: 
            2.00
            $
          </div>
          <div
            class="cart__close"
            data-testid="remove_button"
          >
            ×
          </div>
        </div>
        <div
          class="cart__item"
        >
          <div>
            2
          </div>
          <div
            class="cart__item-title"
          >
            string2
          </div>
          <div
            class="cart__item-price"
          >
            1
            $
          </div>
          <div
            class="cart__item-counter"
          >
            <button
              class="cart__item-counter-btn"
            >
              +
            </button>
            <div
              class="cart__item-counter-count"
              data-testid="item-counter-count"
            >
              2
            </div>
            <button
              class="cart__item-counter-btn minus"
              data-testid="decrease_button"
            >
              -
            </button>
          </div>
          <div
            class="cart__item-price"
            data-testid="total-price-item-counter-count"
          >
            Total: 
            2.00
            $
          </div>
          <div
            class="cart__close"
            data-testid="remove_button"
          >
            ×
          </div>
        </div>
        <div
          class="cart-total-money"
        >
          Сумма за все товары : 
          3
          $
        </div>
        <button
          class="cart__btn"
          disabled=""
        >
          Оплатить
        </button>
        <button
          data-testid="clear_cart_button"
        >
          Очистить корзину
        </button>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
